persistence/gov.go

```diff
@@ -9,893 +14,85 @@ import (
+	return v, err
 }
 
-func (p PostgresContext) GetMessageUnpauseValidatorFeeOwner() ([]byte, error) {
-	return p.GetBytesParam(types.MessageUnpauseValidatorFeeOwner)
+func (p PostgresContext) GetStringParam(paramName string, height int64) (string, error) {
+	v, _, err := getParamOrFlag[string](p, schema.ParamsTableName, paramName, height)
+	return v, err
 }
 
-func (p PostgresContext) GetMessageStakeServiceNodeFeeOwner() ([]byte, error) {
-	return p.GetBytesParam(types.MessageStakeServiceNodeFeeOwner)
+func (p PostgresContext) GetBytesParam(paramName string, height int64) (param []byte, err error) {
+	v, _, err := getParamOrFlag[[]byte](p, schema.ParamsTableName, paramName, height)
+	return v, err
 }
 
-func (p PostgresContext) GetMessageEditStakeServiceNodeFeeOwner() ([]byte, error) {
-	return p.GetBytesParam(types.MessageEditStakeServiceNodeFeeOwner)
+func (p PostgresContext) SetParam(paramName string, value any) error {
+	return p.setParamOrFlag(paramName, value, nil)
 }
 
-func (p PostgresContext) GetMessageUnstakeServiceNodeFeeOwner() ([]byte, error) {
-	return p.GetBytesParam(types.MessageUnstakeServiceNodeFeeOwner)
+func (p PostgresContext) InitFlags() error {
+	// TODO: not implemented
+	return nil
 }
 
-func (p PostgresContext) GetMessagePauseServiceNodeFeeOwner() ([]byte, error) {
-	return p.GetBytesParam(types.MessagePauseServiceNodeFeeOwner)
+func (p PostgresContext) GetIntFlag(flagName string, height int64) (value int, enabled bool, err error) {
+	return getParamOrFlag[int](p, schema.FlagsTableName, flagName, height)
 }
 
-func (p PostgresContext) GetMessageUnpauseServiceNodeFeeOwner() ([]byte, error) {
-	return p.GetBytesParam(types.MessageUnpauseServiceNodeFeeOwner)
+func (p PostgresContext) GetStringFlag(flagName string, height int64) (value string, enabled bool, err error) {
+	return getParamOrFlag[string](p, schema.FlagsTableName, flagName, height)
 }
 
-func (p PostgresContext) GetMessageChangeParameterFeeOwner() ([]byte, error) {
-	return p.GetBytesParam(types.MessageChangeParameterFeeOwner)
+func (p PostgresContext) GetBytesFlag(flagName string, height int64) (value []byte, enabled bool, err error) {
+	return getParamOrFlag[[]byte](p, schema.FlagsTableName, flagName, height)
 }
 
-func (p PostgresContext) GetServiceNodesPerSessionAt(height int64) (int, error) {
-	return p.GetIntParam(types.ServiceNodesPerSessionParamName)
+func (p PostgresContext) SetFlag(flagName string, value any, enabled bool) error {
+	return p.setParamOrFlag(flagName, value, &enabled)
 }
 
-func (p PostgresContext) InitParams() error {
-	ctx, conn, err := p.GetCtxAndConnection()
-	if err != nil {
-		return err
+// setParamOrFlag simply wraps the call to the generic function with the supplied underlying type
+func (p PostgresContext) setParamOrFlag(name string, value any, enabled *bool) error {
+	switch t := value.(type) {
+	case int:
+		return setParamOrFlag(p, name, t, enabled)
```

Comment:
Could we just use ` fallthrough` here through to the last non-default type?

Reply:
This switch statement hurts my eyes too. The best way would have been if we were allowed to use a single case where `t` ended up having the matched type but that's not possible at the moment, Go reverts `t` to type `any` inside the case in that case (pun intended).

I guess Go is forcing the developer to be explicit at the cost of being... verbose.

I played a bit with `fallthrough` but I couldn't use it because of the error `MisplacedFallthrough` and after a quick search I ended up here:

https://github.com/golang/go/issues/40462

It sounds like it's still under discussion.